src/fsharp/TastOps.fs

```diff
@@ -7748,86 +7748,120 @@ let (|RangeInt32Step|_|) g expr =
 
     | _ -> None
 
-let (|ExtractTypeOfExpr|_|) g expr = Some (tyOfExpr g expr)
+let (|GetEnumeratorCall|_|) expr =   
+    match expr with   
+    | Expr.Op (TOp.ILCall( _,  _,  _,  _,  _,  _, _, iLMethodRef,  _,  _,  _),_,[Expr.Val(vref,_,_) | Expr.Op(_, _, [Expr.Val(vref, ValUseFlag.NormalValUse, _)], _) ],_) ->  
+        if iLMethodRef.Name = "GetEnumerator" then Some(vref)  
+        else None  
+    | _ -> None  
+
+let (|CompiledForEachExpr|_|) g expr =   
+    match expr with
+    | Let (enumerableVar, enumerableExpr, _,
+           Let (enumeratorVar, GetEnumeratorCall enumerableVar2, enumeratorBind,
+              TryFinally (WhileLoopForCompiledForEachExpr (_, Let (elemVar,_,_,bodyExpr), _), _))) 
+                 // Apply correctness conditions to ensure this really is a compiled for-each expression.
+                 when valRefEq g (mkLocalValRef enumerableVar) enumerableVar2 &&
+                      enumerableVar.IsCompilerGenerated &&
+                      enumeratorVar.IsCompilerGenerated &&
+                      let fvs = (freeInExpr CollectLocals bodyExpr)
+                      not (Zset.contains enumerableVar fvs.FreeLocals) && 
+                      not (Zset.contains enumeratorVar fvs.FreeLocals) ->
```

Comment:
Yes, I think so - is there something I've missed?  The "bodyExpr" is the body of the while loop, but it shouldn't directly refer to either enumerableVar (the variable holding the enumerable collection, which is consumed in GetEnumeratorCall) or enumeratorVar (the variable holding the enumerator for the collection, which is consumed in the binding for the elemVar "let").